// zenn-cli/dist/server/lib/helper.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.resolveHostname = exports.getLocalIPAddress = exports.completeHtml = exports.generateFileIfNotExist = exports.bufferToDataURL = exports.getImageSize = exports.listFilenamesOrderByModified = exports.listFilenames = exports.listDirnamesOrderByModified = exports.listDirnames = exports.getImageRaw = exports.getFileRaw = exports.getWorkingPath = exports.getPublishedCliVersion = exports.getCurrentCliVersion = exports.generateSlug = void 0;
var crypto_1 = __importDefault(require("crypto"));
var fs_extra_1 = __importDefault(require("fs-extra"));
var path_1 = __importDefault(require("path"));
var Log = __importStar(require("./log"));
var node_fetch_1 = __importDefault(require("node-fetch"));
var cheerio_1 = __importDefault(require("cheerio"));
var url_1 = __importDefault(require("url"));
var os_1 = require("os");
function generateSlug() {
    return crypto_1.default.randomBytes(7).toString('hex');
}
exports.generateSlug = generateSlug;
function getCurrentCliVersion() {
    return require('../../../package.json').version;
}
exports.getCurrentCliVersion = getCurrentCliVersion;
function getPublishedCliVersion() {
    return __awaiter(this, void 0, void 0, function () {
        var response, data, latest;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, node_fetch_1.default("https://registry.npmjs.org/zenn-cli/latest")];
                case 1:
                    response = _a.sent();
                    return [4 /*yield*/, response.json()];
                case 2:
                    data = _a.sent();
                    latest = data['version'];
                    return [2 /*return*/, latest];
            }
        });
    });
}
exports.getPublishedCliVersion = getPublishedCliVersion;
function getWorkingPath(pathFromWorkingDir) {
    // Prevent directory traversal
    if (/\.\./.test(pathFromWorkingDir)) {
        Log.error('取得するファイル/ディレクトリの名前に不正な文字列が含まれているため処理を終了します');
        process.exit(1);
    }
    // remove beginning slash
    return path_1.default.join(process.cwd(), pathFromWorkingDir.replace(/^\//, ''));
}
exports.getWorkingPath = getWorkingPath;
function getFileRaw(fullpath) {
    try {
        var raw = fs_extra_1.default.readFileSync(fullpath, 'utf8');
        return raw;
    }
    catch (err) {
        return null;
    }
}
exports.getFileRaw = getFileRaw;
function getImageRaw(fullpath) {
    try {
        var raw = fs_extra_1.default.readFileSync(fullpath);
        return raw;
    }
    catch (err) {
        return null;
    }
}
exports.getImageRaw = getImageRaw;
function listDirnames(searchDirFullpath) {
    try {
        var allFiles = fs_extra_1.default.readdirSync(searchDirFullpath, {
            withFileTypes: true,
        });
        return allFiles
            .filter(function (file) { return file.isDirectory(); })
            .map(function (_a) {
            var name = _a.name;
            return name;
        });
    }
    catch (e) {
        return null;
    }
}
exports.listDirnames = listDirnames;
function listDirnamesOrderByModified(searchDirFullpath) {
    var allDirnames = listDirnames(searchDirFullpath);
    if (!allDirnames)
        return allDirnames;
    return allDirnames
        .map(function (dirname) {
        return {
            name: dirname,
            time: fs_extra_1.default
                .statSync(path_1.default.join(searchDirFullpath, dirname))
                .ctime.getTime(),
        };
    })
        .sort(function (a, b) { return b.time - a.time; })
        .map(function (_a) {
        var name = _a.name;
        return name;
    });
}
exports.listDirnamesOrderByModified = listDirnamesOrderByModified;
function listFilenames(searchDirFullpath) {
    try {
        var allFiles = fs_extra_1.default.readdirSync(searchDirFullpath);
        return allFiles;
    }
    catch (e) {
        return null;
    }
}
exports.listFilenames = listFilenames;
function listFilenamesOrderByModified(searchDirFullpath) {
    var allFiles = listFilenames(searchDirFullpath);
    if (!allFiles)
        return allFiles;
    return allFiles
        .map(function (filename) {
        return {
            name: filename,
            time: fs_extra_1.default
                .statSync(path_1.default.join(searchDirFullpath, filename))
                .ctime.getTime(),
        };
    })
        .sort(function (a, b) { return b.time - a.time; })
        .map(function (_a) {
        var name = _a.name;
        return name;
    });
}
exports.listFilenamesOrderByModified = listFilenamesOrderByModified;
function getImageSize(fullpath) {
    var stat = fs_extra_1.default.statSync(fullpath);
    return stat.size;
}
exports.getImageSize = getImageSize;
function bufferToDataURL(buffer, mediaType) {
    return "data:" + mediaType + ";base64," + buffer.toString('base64');
}
exports.bufferToDataURL = bufferToDataURL;
function generateFileIfNotExist(fullpath, content) {
    fs_extra_1.default.outputFileSync(fullpath, content, { flag: 'wx' } // Don't overwrite
    );
}
exports.generateFileIfNotExist = generateFileIfNotExist;
function completeHtml(html) {
    var $ = cheerio_1.default.load(html);
    $('img').map(function (i, el) {
        var src = el.attribs['src'];
        // srcがURLの場合はチェックしない
        if (isUrl(src))
            return;
        // 先頭が `/images/` であること
        if (!path_1.default.isAbsolute(src)) {
            $(el).before("<p style=\"color: var(--c-error); font-weight: 700\"><code>" + src + "</code>\u3092\u8868\u793A\u3067\u304D\u307E\u305B\u3093\u3002\u30ED\u30FC\u30AB\u30EB\u306E\u753B\u50CF\u3092\u8AAD\u307F\u8FBC\u3080\u306B\u306F\u76F8\u5BFE\u30D1\u30B9\u3067\u306F\u306A\u304F<code>/images/example.png</code>\u306E\u3088\u3046\u306B<code>/images/</code>\u304B\u3089\u59CB\u307E\u308B\u30D1\u30B9\u3092\u6307\u5B9A\u3057\u3066\u304F\u3060\u3055\u3044\u3002</p>");
            $(el).remove();
            return;
        }
        // 拡張子が png,jpg,jpeg,gif であること
        if (!src.match(/(.png|.jpg|.jpeg|.gif)$/)) {
            $(el).before("<p style=\"color: var(--c-error); font-weight: 700\"><code>" + src + "</code>\u3092\u8868\u793A\u3067\u304D\u307E\u305B\u3093\u3002\u5BFE\u5FDC\u3057\u3066\u3044\u308B\u753B\u50CF\u306E\u62E1\u5F35\u5B50\u306F <code>png, jpg, jpeg, gif</code> \u3067\u3059\u3002</p>");
            $(el).remove();
            return;
        }
        var filepath = getWorkingPath(src);
        if (!fs_extra_1.default.existsSync(filepath)) {
            $(el).before("<p style=\"color: var(--c-error); font-weight: 700\"><code>" + src + "</code>\u306B\u30D5\u30A1\u30A4\u30EB\u304C\u5B58\u5728\u3057\u307E\u305B\u3093\u3002</p>");
            $(el).remove();
            return;
        }
        var fileSize = fs_extra_1.default.statSync(filepath).size;
        if (fileSize > 1024 * 1024 * 3) {
            $(el).before("<p style=\"color: var(--c-error); font-weight: 700\"><code>" + src + "</code>\u306E\u30D5\u30A1\u30A4\u30EB\u30B5\u30A4\u30BA\uFF08" + Math.trunc((fileSize * 100) / 1024 / 1024) / 100 + " MB\uFF09\u306F\u30A2\u30C3\u30D7\u30ED\u30FC\u30C9\u53EF\u80FD\u306A\u30B5\u30A4\u30BA\u3092\u8D85\u3048\u3066\u3044\u307E\u3059\u3002\u30D5\u30A1\u30A4\u30EB\u30B5\u30A4\u30BA\u306F3MB\u4EE5\u5185\u306B\u3057\u3066\u304F\u3060\u3055\u3044\u3002</p>");
            $(el).remove();
            return;
        }
    });
    return $.html();
}
exports.completeHtml = completeHtml;
function isUrl(text) {
    try {
        return new url_1.default.URL(text) && true;
    }
    catch (_a) {
        return false;
    }
}
var getLocalIPAddress = function () {
    var interfaces = os_1.networkInterfaces();
    var networks = Object.values(interfaces).flatMap(function (v) { return v !== null && v !== void 0 ? v : []; });
    for (var _i = 0, networks_1 = networks; _i < networks_1.length; _i++) {
        var detail = networks_1[_i];
        if (detail.family === 'IPv4' && detail.address !== '127.0.0.1') {
            return detail.address;
        }
    }
};
exports.getLocalIPAddress = getLocalIPAddress;
var resolveHostname = function (hostname) {
    var name = 'localhost';
    var host = hostname;
    if (hostname) {
        if (hostname !== '0.0.0.0' && hostname !== '::') {
            name = hostname;
        }
        else {
            host = exports.getLocalIPAddress();
        }
    }
    return { name: name, host: host };
};
exports.resolveHostname = resolveHostname;
