// zenn-cli/dist/client/assets/index.6438bf0f.js
var e=Object.defineProperty,t=Object.getOwnPropertySymbols,r=Object.prototype.hasOwnProperty,a=Object.prototype.propertyIsEnumerable,l=(t,r,a)=>r in t?e(t,r,{enumerable:!0,configurable:!0,writable:!0,value:a}):t[r]=a,i=(e,i)=>{for(var n in i||(i={}))r.call(i,n)&&l(e,n,i[n]);if(t)for(var n of t(i))a.call(i,n)&&l(e,n,i[n]);return e};import{q as n,R as o,r as s,u as c,a as m,d,b as p,c as u,W as g,t as h,e as _,f as E,g as b,h as v,i as f,j as y}from"./vendor.ec9681ae.js";let x;const k={},w=e=>o.createElement(N,{className:"error-message"},o.createElement("div",{className:"error-message__inner"},o.createElement("img",{src:"https://twemoji.maxcdn.com/2/svg/1f63f.svg",alt:"",width:90,height:90,className:"error-message__icon"}),o.createElement("div",{className:"error-message__text"},e.message||"ã‚¨ãƒ©ãƒ¼ãŒç™ºç”Ÿã—ã¾ã—ãŸ"))),N=n.div`
  height: 100vh;
  display: flex;
  align-items: center;
  justify-content: center;
  .error-message__inner {
    transform: translateY(-25%);
    padding: 1rem;
  }
  .error-message__icon {
    display: block;
    margin: 0 auto;
  }
  .error-message__text {
    margin-top: 1.7rem;
    text-align: center;
    font-weight: 700;
    font-size: 1.4rem;
  }
`;class $ extends o.Component{constructor(){super(...arguments),this.state={error:null}}componentDidCatch(e){const t=s.isLocationNotFoundError(e)?"NotFound":"Unknown";this.setState({error:t})}render(){if(null===this.state.error)return this.props.children;const e="NotFound"===this.state.error?"ãƒšãƒ¼ã‚¸ãŒè¦‹ã¤ã‹ã‚Šã¾ã›ã‚“ã§ã—ãŸ":"åŽŸå› ä¸æ˜Žã®ã‚¨ãƒ©ãƒ¼ãŒç™ºç”Ÿã—ã¾ã—ãŸ";return o.createElement(w,{message:e})}}async function C(e){const t=await fetch(e);if(!t.ok){const e=new Error("An error occurred while fetching the data."),r=await t.json();throw e.message=r.message,e}return t.json()}function z(e,t){return c(e,C,t)}const F=m.exports.createContext({reloadedAt:0}),j=e=>{const[t,r]=m.exports.useState(0);return m.exports.useEffect((()=>{const e="https:"===window.location.protocol?"wss:":"ws:",t=new WebSocket(`${e}//${window.location.host}`);return t.onmessage=()=>{r((new Date).getTime())},()=>{console.log("Disconnecting socket..."),t.close()}}),[]),o.createElement(F.Provider,{value:{reloadedAt:t}},e.children)};function A(e){const{reloadedAt:t}=m.exports.useContext(F);m.exports.useEffect((()=>{0!==t&&e()}),[t])}function M({cacheKey:e,defaultValue:t}){const r=m.exports.useMemo((()=>{const r=function(e){const t=localStorage.getItem(e);if(!t)return null;const r=JSON.parse(t);if("object"!=typeof r||null===r)return function(e){localStorage.removeItem(e)}(e),null;return r.value}(e);return null===r?t:typeof r!=typeof t?(console.log(`Persisted value type(${typeof r}) is invalid. It must be ${typeof t}`),localStorage.removeItem(e),t):r}),[]),[a,l]=m.exports.useState(r);return[a,function(t){l(t),function(e,t){const r={value:t},a=JSON.stringify(r);localStorage.setItem(e,a)}(e,t)}]}const V=e=>o.createElement(S,{className:"list-item-inner"},"string"==typeof e.emoji&&o.createElement("span",{className:"list-item-inner__emoji"},e.emoji),"string"==typeof e.iconImgSrc&&o.createElement("img",{src:e.iconImgSrc,width:18,height:18,alt:"",className:"list-item-inner__icon"}),!!e.label&&o.createElement("span",{className:"list-item-inner__label"},e.label),o.createElement("div",{className:"list-item-inner__title",title:e.title},e.title,!0===e.showNewTabIcon&&o.createElement(d,{className:"list-item-inner__icon-open"}))),S=n.div`
  display: flex;
  align-items: center;

  .list-item-inner__emoji {
    margin-right: 7px;
  }
  .list-item-inner__icon {
    margin-right: 5px;
  }
  .list-item-inner__label {
    display: inline-block;
    padding: 1px 3px;
    margin-right: 4px;
    border: solid 1px var(--c-gray-border);
    color: var(--c-gray);
    font-size: 11px;
    border-radius: 4px;
    background: #fff;
    line-height: 1.3;
  }
  .list-item-inner__title {
    flex: 1;
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
  }
  .list-item-inner__icon-open {
    margin: 0 2px;
    width: 15px;
    height: 15px;
    vertical-align: -2px;
    opacity: 0.7;
  }
`,O=e=>{const t=`toggle-${e.uniqueKey}`,r=`toggle-${e.uniqueKey}-open`,[a,l]=M({cacheKey:r,defaultValue:e.defaultOpen});return o.createElement(R,{className:"directory"},o.createElement("button",{className:"directory__button",type:"button",role:"tab","aria-controls":t,"aria-expanded":a,"aria-label":"ãƒ•ã‚©ãƒ«ãƒ€ãƒ¼ã‚’é–‹ã",onClick:()=>l(!a)},o.createElement(p,{className:"directory__icon-arrow "+(a?"rotated":"")}),o.createElement(V,{title:e.title,label:e.label,iconImgSrc:"/static-images/"+(a?"folder-open.svg":"folder-close.svg")})),o.createElement("div",{className:"directory__children",role:"tabpanel","aria-hidden":!a,id:t},a?e.children:null))},R=n.div`
  .directory__button {
    position: relative;
    padding: 5px 0 5px 15px;
    display: block;
    width: 100%;
    color: var(--c-gray);
    text-align: left;
    &:hover {
      color: var(--c-body);
    }
  }
  .directory__icon-arrow {
    position: absolute;
    width: 12px;
    height: 12px;
    top: 8px;
    left: -2px;
    color: var(--c-gray);
    &.rotated {
      transform: rotate(90deg);
    }
  }

  .directory__children {
    padding-left: 15px;
  }
`,L=e=>{const[t,r]=m.exports.useState(!1);return o.createElement(B,{className:"settings"},o.createElement("button",{className:"settings__open",onClick:()=>r(!0)},e.openButtonIcon),t&&o.createElement(o.Fragment,null,o.createElement("div",{className:"settings__backdrop",onClick:()=>r(!1)}),o.createElement("div",{className:`settings__popover settings__popover--${e.position}`,style:{width:e.width}},e.options.map((({value:t,label:a})=>o.createElement("button",{key:`option-${t}`,value:t,onClick:()=>function(t){e.setValue(t),setTimeout((()=>{r(!1)}),200)}(t),className:"settings__option","aria-selected":t===e.value},o.createElement(u,{className:"settings__option-icon"}),a))))))},B=n.div`
  display: flex;
  align-items: center;
  position: relative;

  .settings__open {
    display: inline-flex;
    align-items: center;
    color: var(--c-gray);
    &:hover {
      color: var(--c-body);
    }
  }
  .settings__backdrop {
    position: fixed;
    background: transparent;
    top: 0;
    left: 0;
    bottom: 0;
    right: 0;
  }
  .settings__backdrop,
  .settings__popover {
    z-index: 9999;
  }
  .settings__popover {
    position: absolute;
    background: #fff;
    top: calc(100% + 8px);
    box-shadow: 0 2px 8px rgb(15 26 62 / 35%);
    border-radius: 5px;
    font-size: 13.5px;
    overflow: hidden;
    padding: 5px;
  }
  .settings__popover--left {
    left: -10px;
  }
  .settings__popover--right {
    right: -10px;
  }
  .settings__popover--center {
    left: 50%;
    transform: translateX(-50%);
  }
  .settings__option {
    width: 100%;
    text-align: left;
    padding: 8px 8px 8px 22px;
    border-radius: 4px;
    &:hover {
      background: var(--c-gray-bg);
    }
    &[aria-selected='true'] {
      .settings__option-icon {
        opacity: 1;
      }
    }
  }
  .settings__option-icon {
    position: absolute;
    left: 8px;
    width: 16px;
    height: 16px;
    opacity: 0;
    color: var(--c-primary);
  }
`,I=e=>{var t;return"rawHtml"in e?(null==(t=e.rawHtml)?void 0:t.length)?o.createElement("div",{className:"znc",dangerouslySetInnerHTML:{__html:e.rawHtml}}):o.createElement(P,null,"æœ¬æ–‡ã‚’å…¥åŠ›ã—ã¦ãã ã•ã„"):o.createElement("div",{className:"znc"},e.children)},P=n.div`
  text-align: center;
  font-size: 1.1rem;
  color: var(--c-gray);
`,U=n.div`
  max-width: 760px;
  margin: 0 auto;
  padding: 0 1.8rem;
  @media (max-width: 468px) {
    padding: 0 1rem;
  }
`;function H(e){m.exports.useEffect((()=>{e&&(document.title=e)}),[e])}const T=()=>{const{data:e}=z("/api/cli-version",{revalidateOnFocus:!1,revalidateOnReconnect:!1});return e?!1===e.updateAvailable?null:o.createElement(D,{className:"update-notification-bar"},o.createElement(U,null,o.createElement("div",{className:"update-notification-bar__inner"},o.createElement("div",{className:"update-notification-bar__icon"},"!"),o.createElement("div",{className:"update-notification-bar__text"},"æ–°ã—ã„ãƒãƒ¼ã‚¸ãƒ§ãƒ³ãŒãƒªãƒªãƒ¼ã‚¹ã•ã‚Œã¦ã„ã¾ã™ã€‚",o.createElement("code",null,"npm install zenn-cli@latest"),"ã§æ›´æ–°ã—ã¦ãã ã•ã„")))):null},K=g`
  from {
   opacity: 0;
   transform: translateY(-40px);
  }
  to {
    opacity: 1;
    transform: translateY(0);
  }
`,D=n.div`
  .update-notification-bar__inner {
    display: flex;
    justify-content: center;
    align-items: center;
    flex-wrap: wrap;
    margin-top: 1.3rem;
    padding: 0.9rem;
    background: var(--c-primary-bg);
    color: #364757;
    border-radius: 10px;
    animation: ${K} 0.3s ease;
  }

  .update-notification-bar__icon {
    margin-right: 8px;
    content: '!';
    font-weight: 700;
    width: 20px;
    line-height: 20px;
    height: 20px;
    text-align: center;
    border-radius: 50%;
    background: var(--c-primary);
    color: #fff;
    font-size: 12px;
  }
  .update-notification-bar__text {
    flex: 1;
    font-size: 14px;
    code {
      display: inline-flex;
      margin: 0 0.2em;
      padding: 0.1em 0.6em;
      font-size: 0.9em;
      background: #cfe6f9;
      border-radius: 5px;
    }
  }
`,q=()=>{const{data:e}=z("/api/local-info"),t=null==e?void 0:e.hasInit;return H("Zenn Editor"),o.createElement(o.Fragment,null,o.createElement(T,null),o.createElement(U,null,o.createElement(W,{className:"home"},o.createElement("h1",{className:"home__title"},o.createElement("img",{src:"/logo.svg",alt:"Zenn Editor",width:300,height:35})),o.createElement("div",{className:"home__content"},o.createElement(I,null,!0===t||void 0===t?o.createElement(o.Fragment,null,o.createElement("h3",null,"ðŸ“ è¨˜äº‹ã®ä½œæˆ"),o.createElement("pre",null,o.createElement("code",null,"$ npx zenn"," ",o.createElement("span",{className:"token function"},"new:article"),o.createElement("div",{className:"token comment"},"# è¨˜äº‹ã®URLã®ä¸€éƒ¨ã¨ãªã‚‹slugã‚’æŒ‡å®šã—ã¦ä½œæˆã™ã‚‹ã“ã¨ã‚‚ã§ãã¾ã™ã€‚",o.createElement("br",null),"$ npx zenn new:article --slug my-awesome-article"))),o.createElement("h3",null,"ðŸ“˜ æœ¬ã®ä½œæˆ"),o.createElement("pre",null,o.createElement("code",null,"$ npx zenn"," ",o.createElement("span",{className:"token function"},"new:book"),o.createElement("div",{className:"token comment"},"# æœ¬ã®URLã®ä¸€éƒ¨ã¨ãªã‚‹slugã‚’æŒ‡å®šã—ã¦ä½œæˆã™ã‚‹ã“ã¨ã‚‚ã§ãã¾ã™ã€‚",o.createElement("br",null),"$ npx zenn new:book --slug my-awesome-book")))):o.createElement(o.Fragment,null,o.createElement("h3",null,"ðŸ”§ ã‚»ãƒƒãƒˆã‚¢ãƒƒãƒ—ãŒå¿…è¦ã§ã™"),o.createElement("p",null,"ã¾ã ",o.createElement("code",null,"articles"),"ãƒ‡ã‚£ãƒ¬ã‚¯ãƒˆãƒªãŒä½œæˆã•ã‚Œã¦ã„ã¾ã›ã‚“ã€‚ä»¥ä¸‹ã®ã‚³ãƒžãƒ³ãƒ‰ã‚’å®Ÿè¡Œã—ã¦ã‚»ãƒƒãƒˆã‚¢ãƒƒãƒ—ã‚’è¡Œã£ã¦ãã ã•ã„ã€‚"),o.createElement("pre",null,o.createElement("code",null,"$ npx zenn ",o.createElement("span",{className:"token function"},"init"))))),o.createElement("div",{className:"home__learn-more"},o.createElement("div",{className:"home__learn-more-title"},"è©³ã—ã„ä½¿ã„æ–¹ã¯CLIãƒªãƒ•ã‚¡ãƒ¬ãƒ³ã‚¹ã‚’ã”è¦§ãã ã•ã„"),o.createElement(lt,{slug:"zenn-cli-guide",className:"home__link-cli-guide"},"CLIãƒªãƒ•ã‚¡ãƒ¬ãƒ³ã‚¹ã‚’é–‹ã"))))))},W=n.article`
  .home__title {
    padding: 2.5rem 0;
    text-align: center;
    border-bottom: solid 1px var(--c-gray-border);
  }
  .home__content {
    padding: 2.5rem 0;
  }
  .home__learn-more {
    margin-top: 3rem;
    padding: 1.8rem 2rem 2.5rem;
    background: var(--c-gray-bg);
    border-radius: 10px;
    text-align: center;
  }
  .home__learn-more-title {
    color: var(--c-gray);
  }
  .home__link-cli-guide {
    margin-top: 1rem;
    display: inline-flex;
    padding: 0.6em 1.5em;
    background: var(--c-primary);
    color: #fff;
    border-radius: 0.45em;
    border: solid 1px rgba(92, 147, 187, 0.15);
    font-size: 0.95rem;
    font-weight: 600;
  }
`,Y=e=>o.createElement(X,{style:{margin:e.margin}}),Z=g`
  to {
    transform: rotate(360deg);
  }
`,J=g`
  from {
    opacity: 0;
  }
  to {
    opacity: 1;
  }
`,X=n.div`
  display: table;
  width: 30px;
  height: 30px;
  margin: 0 auto;
  border: 4px solid var(--c-primary);
  border-radius: 50%;
  border-top-color: var(--c-primary-bg);
  animation: ${Z} 0.8s linear infinite, ${J} 0.7s;
`,G=({hash:e,slug:t})=>{const{data:r,error:a}=z(`/api/cli-guide/${t}`,{revalidateOnFocus:!1,revalidateOnReconnect:!1,errorRetryCount:2}),l=null==r?void 0:r.bodyHtml,i=null==r?void 0:r.title;return H(i),m.exports.useEffect((()=>{if(!e||!l)return;const t=document.getElementById(e);t?(t.scrollIntoView(),window.scrollBy(0,-30)):console.error(`Couldn't find elements with id="${e}"`)}),[l,e]),a?o.createElement("div",null,a.message):l?o.createElement(o.Fragment,null,o.createElement(U,null,o.createElement(Q,{className:"guide"},o.createElement("h1",{className:"guide__title"},i),o.createElement("div",{className:"guide__content"},o.createElement(I,{rawHtml:l}))))):o.createElement(Y,{margin:"5rem auto"})},Q=n.div`
  padding: 3rem 0;
  font-size: 0.94rem;
  .guide__title {
    font-size: 2rem;
  }
  .guide__content {
    padding: 1.5rem 0;
  }
`;function ee(e){return"string"==typeof e&&e.length>0}const te={isCritical:!0,getMessage:({slug:e})=>function(e){return`slugã®å€¤ï¼ˆ${e}ï¼‰ãŒä¸æ­£ã§ã™ã€‚å°æ–‡å­—ã®åŠè§’è‹±æ•°å­—ï¼ˆa-z0-9ï¼‰ã€ãƒã‚¤ãƒ•ãƒ³ï¼ˆ-ï¼‰ã€ã‚¢ãƒ³ãƒ€ãƒ¼ã‚¹ã‚³ã‚¢ï¼ˆ_ï¼‰ã®12ã€œ50å­—ã®çµ„ã¿åˆã‚ã›ã«ã—ã¦ãã ã•ã„`}(e),isValid:({slug:e})=>function(e){return!!e&&/^[0-9a-z\-_]{12,50}$/.test(e)}(e)},re={isCritical:!0,getMessage:()=>"titleï¼ˆã‚¿ã‚¤ãƒˆãƒ«ï¼‰ã‚’æ–‡å­—åˆ—ã§å…¥åŠ›ã—ã¦ãã ã•ã„",isValid:({title:e})=>ee(e)},ae={isCritical:!0,getMessage:()=>"ã‚¿ã‚¤ãƒˆãƒ«ã¯70å­—ä»¥å†…ã«ã—ã¦ãã ã•ã„",isValid:({title:e})=>!ee(e)||e.length<=70},le={isCritical:!0,getMessage:()=>'publishedï¼ˆå…¬é–‹è¨­å®šï¼‰ã‚’ true ã‹ false ã§æŒ‡å®šã—ã¦ãã ã•ã„ï¼ˆã‚¯ã‚ªãƒ†ãƒ¼ã‚·ãƒ§ãƒ³ " ã§å›²ã¾ãªã„ã§ãã ã•ã„ï¼‰',isValid:({published:e})=>"boolean"==typeof e},ie={isCritical:!0,detailUrl:"https://zenn.dev/tech-or-idea",detailUrlText:"è©³ç´°ã‚’é–‹ã",getMessage:()=>"typeï¼ˆè¨˜äº‹ã®ã‚¿ã‚¤ãƒ—ï¼‰ã« tech ã‚‚ã—ãã¯ idea ã‚’æŒ‡å®šã—ã¦ãã ã•ã„ã€‚æŠ€è¡“è¨˜äº‹ã®å ´åˆã¯ tech ã‚’æŒ‡å®šã—ã¦ãã ã•ã„",isValid:({type:e})=>!!e&&("tech"===e||"idea"===e)},ne={detailUrl:"https://getemoji.com",detailUrlText:"çµµæ–‡å­—ã‚’æŽ¢ã™",getMessage:()=>"ã‚¢ã‚¤ã‚­ãƒ£ãƒƒãƒã¨ãªã‚‹ emojiï¼ˆçµµæ–‡å­—ï¼‰ã‚’æŒ‡å®šã—ã¦ãã ã•ã„",isValid:({emoji:e})=>ee(e)},oe={isCritical:!0,getMessage:()=>"çµµæ–‡å­—ï¼ˆemojiï¼‰ã‚’1ã¤ã ã‘æŒ‡å®šã—ã¦ãã ã•ã„",isValid:({emoji:e})=>{if(!ee(e))return!0;const t=h(),r=e.match(t);return!(!r||!r[0]||r[1])}},se={isCritical:!0,getMessage:()=>'topicsï¼ˆè¨˜äº‹ã«é–¢é€£ã™ã‚‹è¨€èªžã‚„æŠ€è¡“ï¼‰ã‚’é…åˆ—ã§æŒ‡å®šã—ã¦ãã ã•ã„ã€‚ä¾‹ï¼‰["react", "javascript"]',isValid:({topics:e})=>!!Array.isArray(e)&&e.length>0},ce={isCritical:!0,getMessage:()=>"topicsã¯æœ€å¤§5ã¤ã¾ã§æŒ‡å®šã§ãã¾ã™",isValid:({topics:e})=>!Array.isArray(e)||e.length<=5},me={isCritical:!0,getMessage:()=>"topicsã¯å…¨ã¦æ–‡å­—åˆ—ã§æŒ‡å®šã—ã¦ãã ã•ã„",isValid:({topics:e})=>!Array.isArray(e)||e.every((e=>"string"==typeof e&&e.length>0))},de={getMessage:()=>"topicsã«è¨˜å·ã‚„ã‚¹ãƒšãƒ¼ã‚¹ã‚’ä½¿ç”¨ã™ã‚‹ã“ã¨ã¯ã§ãã¾ã›ã‚“ã€‚ä¾‹ãˆã°C++ã¯ã€Œcppã€ã€C#ã¯ã€Œcsharpã€ã¨è¨˜è¼‰ã—ã¦ãã ã•ã„",isValid:({topics:e})=>{if(!Array.isArray(e))return!0;return!e.some((e=>"string"==typeof e&&e.match(/[ -/:-@[-`{-~]/g)))}},pe={getMessage:()=>"tagsã§ã¯ãªãtopicsã‚’ä½¿ã£ã¦ãã ã•ã„",isValid:e=>{var t,r;return!(null==(t=e.tags)?void 0:t.length)&&!(null==(r=e.tag)?void 0:r.length)}},ue={isCritical:!0,getMessage:()=>"summaryï¼ˆæœ¬ã®èª¬æ˜Žï¼‰ã®è¨˜è¼‰ã¯å¿…é ˆã§ã™",isValid:({summary:e})=>"string"==typeof e&&e.length>0},ge={isCritical:!0,getMessage:()=>'priceï¼ˆæœ¬ã®ä¾¡æ ¼ï¼‰ã‚’åŠè§’æ•°å­—ã§æŒ‡å®šã—ã¦ãã ã•ã„ï¼ˆã‚¯ã‚ªãƒ†ãƒ¼ã‚·ãƒ§ãƒ³ " ã§å›²ã¾ãªã„ã§ãã ã•ã„ï¼‰',isValid:({price:e})=>"number"==typeof e},he={isCritical:!0,getMessage:()=>"priceï¼ˆæœ¬ã®ä¾¡æ ¼ï¼‰ã‚’æœ‰æ–™ã«ã™ã‚‹å ´åˆã€200ã€œ5000ã®é–“ã§æŒ‡å®šã—ã¦ãã ã•ã„",isValid:({price:e})=>"number"!=typeof e||(0===e||e>=200&&e<=5e3)},_e={isCritical:!0,getMessage:()=>"priceï¼ˆæœ¬ã®ä¾¡æ ¼ï¼‰ã¯100å††å˜ä½ã§æŒ‡å®šã—ã¦ãã ã•ã„",isValid:({price:e})=>"number"!=typeof e||e%100==0},Ee={getMessage:({slug:e})=>`æœ¬ã®ã‚«ãƒãƒ¼ç”»åƒï¼ˆcover.pngã‚‚ã—ãã¯cover.jpgï¼‰ã‚’ã€Œ/books/${e}ã€ãƒ‡ã‚£ãƒ¬ã‚¯ãƒˆãƒªã«é…ç½®ã—ã¦ãã ã•ã„`,isValid:({coverDataUrl:e})=>"string"==typeof e},be={isCritical:!0,getMessage:e=>"ã‚«ãƒãƒ¼ç”»åƒã®ã‚µã‚¤ã‚ºã¯1MBä»¥å†…ã«ã—ã¦ãã ã•ã„ã€‚ç¾åœ¨ã®ã‚µã‚¤ã‚º: "+(e.coverFilesize?`${Math.trunc(e.coverFilesize/1024)}KB`:"å–å¾—ã§ãã¾ã›ã‚“ã§ã—ãŸ"),isValid:({coverDataUrl:e,coverFilesize:t})=>"string"!=typeof e||"number"==typeof t&&t<=1048576};function ve(e,t){return t&&e?Math.round(t/e*10)/10:null}const fe={getMessage:e=>{const t=ve(e.coverWidth,e.coverHeight);return"ã‚«ãƒãƒ¼ç”»åƒã®ã€Œå¹… : é«˜ã•ã€ã®æ¯”çŽ‡ã¯ã€Œ1 : 1.4ã€ã«ã™ã‚‹ã“ã¨ã‚’ãŠã™ã™ã‚ã—ã¾ã™ï¼ˆæœ€çµ‚çš„ã«å¹…500pxãƒ»é«˜ã•700pxã«ãƒªã‚µã‚¤ã‚ºã•ã‚Œã¾ã™ï¼‰ã€‚"+(t?`ç¾åœ¨ã®æ¯”çŽ‡ã¯ã€Œ1 : ${t}ã€ã§ã™`:"")},isValid:({coverDataUrl:e,coverHeight:t,coverWidth:r})=>{if("string"!=typeof e)return!0;const a=ve(r,t);return"number"==typeof a&&1.4===a}},ye={isCritical:!0,getMessage:()=>"config.yamlã® chapters ã®æŒ‡å®šã«èª¤ã‚ŠãŒã‚ã‚Šã¾ã™",detailUrl:"https://zenn.dev/zenn/articles/zenn-cli-guide#%F0%9F%93%84-config.yaml",isValid:({specifiedChapterSlugs:e})=>void 0===e||Array.isArray(e)&&e.every((e=>"string"==typeof e))},xe={isCritical:!0,getMessage:()=>"chapters ã«æŒ‡å®šã™ã‚‹æ–‡å­—åˆ—ã«ã¯æ‹¡å¼µå­ï¼ˆ.mdï¼‰ã‚’å«ã‚ãªã„ã§ãã ã•ã„ã€‚ãƒ•ã‚¡ã‚¤ãƒ«åãŒ example.md ã§ã‚ã‚Œã° example ã¨ã®ã¿è¨˜è¼‰ã—ã¦ãã ã•ã„",detailUrl:"https://zenn.dev/zenn/articles/zenn-cli-guide#%F0%9F%93%84-config.yaml",isValid:({specifiedChapterSlugs:e})=>{if(!Array.isArray(e))return!0;return!e.some((e=>"string"==typeof e&&e.match(/\.md$/)))}},ke={isCritical:!0,getMessage:({slug:e})=>function(e){return`ãƒãƒ£ãƒ—ã‚¿ãƒ¼ã®slugã®å€¤ï¼ˆ${e}ï¼‰ãŒä¸æ­£ã§ã™ã€‚å°æ–‡å­—ã®åŠè§’è‹±æ•°å­—ï¼ˆa-z0-9ï¼‰ã€ãƒã‚¤ãƒ•ãƒ³ï¼ˆ-ï¼‰ã€ã‚¢ãƒ³ãƒ€ãƒ¼ã‚¹ã‚³ã‚¢ï¼ˆ_ï¼‰ã®1ã€œ50å­—ã®çµ„ã¿åˆã‚ã›ã«ã—ã¦ãã ã•ã„`}(e),isValid:({slug:e})=>function(e){return!!e&&(/^[0-9a-z\-_]{1,50}$/.test(e)||/^[0-9]+.[0-9a-z\-_]{1,50}$/.test(e))}(e)},we={isCritical:!0,getMessage:()=>"freeï¼ˆç„¡æ–™å…¬é–‹è¨­å®šï¼‰ã«ã¯ true ã‚‚ã—ãã¯ falseã®ã¿ã‚’æŒ‡å®šã—ã¦ãã ã•ã„",isValid:({free:e})=>void 0===e||"boolean"==typeof e};function Ne(e,t){return t.reduce(((t,r)=>(r.isValid(e)||t.push({isCritical:!0===r.isCritical,message:r.getMessage(e),detailUrl:r.detailUrl,detailUrlText:r.detailUrlText}),t)),[])}const $e=({topics:e})=>o.createElement(Ce,{className:"topic-list"},e.map(((e,t)=>o.createElement("span",{className:"topic-list__item",key:`topic-${t}`},e)))),Ce=n.div`
  margin-top: -${"3px"};
  .topic-list__item {
    display: inline-flex;
    color: var(--c-gray);
    font-size: 0.9em;
    border: solid 1px var(--c-gray-border);
    background: #fff;
    padding: 0.1em 0.4em;
    border-radius: 5px;
    margin: ${"3px"} 6px ${"3px"} 0;
  }
`,ze=e=>o.createElement(Fe,{className:"property-row"},o.createElement("div",{className:"property-row__title"},e.title),o.createElement("div",{className:"property-row__content"},e.children)),Fe=n.div`
  display: flex;
  font-size: 0.92rem;
  color: var(--c-gray);
  & + .property-row {
    margin-top: 0.5rem;
  }
  .property-row__title {
    font-weight: 700;
    width: 140px;
  }
  .property-row__content {
    flex: 1;
  }
`,je=({message:e,isCritical:t,detailUrl:r,detailUrlText:a})=>o.createElement(Ae,{className:"validation-error-row "+(t?"validation-error-row--critical":"")},o.createElement(_,{className:"validation-error-row__icon"}),o.createElement("div",{className:"validation-error-row__message"},e,"string"==typeof r&&o.createElement("a",{className:"validation-error-row__link",href:r,target:"_blank",rel:"nofollow noreferrer"},a||"è©³ç´°ã‚’è¦‹ã‚‹",o.createElement(d,{className:"validation-error-row__link-icon"})))),Ae=n.div`
  display: flex;
  align-items: flex-start;
  margin-top: 0.5rem;
  .validation-error-row__icon {
    width: 1.1em;
    height: 1.1em;
    color: ${"#ff9715"};
    flex-shrink: 0;
    transform: translateY(-0.1em);
  }
  .validation-error-row__message {
    flex: 1;
    color: var(--c-gray);
    font-size: 14px;
    margin-left: 7px;
  }
  .validation-error-row__link {
    margin-left: 6px;
    color: inherit;
    text-decoration: underline;
    text-underline-offset: 2px;
    text-decoration-color: var(--c-gray-border);
    display: inline-flex;
    align-items: center;
    &:hover {
      color: var(--c-body);
    }
  }
  .validation-error-row__link-icon {
    width: 15px;
    height: 15px;
  }
  &.validation-error-row--critical {
    .validation-error-row__message {
      /* color: var(--c-error); */
    }
    .validation-error-row__icon {
      color: var(--c-error);
    }
  }
`,Me=({validationErrors:e})=>{const{criticalErrors:t,warnings:r}=m.exports.useMemo((()=>e.reduce(((e,t)=>(e[t.isCritical?"criticalErrors":"warnings"].push(t),e)),{criticalErrors:[],warnings:[]})),[e]);return e.length?o.createElement(Ve,{className:"validation-error"},t.length>0&&o.createElement("div",{className:"validation-error__critical"},o.createElement("div",{className:"validation-error__title validation-error__title--critical"},t.length,"ä»¶ã®ã‚¨ãƒ©ãƒ¼ãŒã‚ã‚Šã¾ã™"),t.map(((e,t)=>o.createElement(je,i({key:`validation-error-${t}`},e))))),r.length>0&&o.createElement("div",{className:"validation-error__warnings"},o.createElement("div",{className:"validation-error__title"},r.length,"ä»¶ã®ä¿®æ­£ã‚’ãŠã™ã™ã‚ã—ã¾ã™"),r.map(((e,t)=>o.createElement(je,i({key:`validation-error-${t}`},e)))))):null},Ve=n.div`
  padding: 1.2rem 1.4rem 1.3rem;
  background: #fff;
  border-radius: 10px;
  .validation-error__title {
    font-size: 0.95rem;
    color: ${"#ff9715"};
    font-weight: 700;
  }
  .validation-error__title--critical {
    color: var(--c-error);
  }
  .validation-error__critical + .validation-error__warnings {
    margin-top: 1.2rem;
  }
`,Se=({article:e})=>{const t=m.exports.useMemo((()=>(e=>Ne(e,[te,re,ae,le,ie,oe,ne,se,pe,de,ce,me]))(e)),[e]);return o.createElement(Oe,null,o.createElement(U,null,"string"==typeof e.emoji&&o.createElement("div",{className:"article-header__emoji"},e.emoji),o.createElement("h1",{className:"article-header__title"},e.title||"titleã‚’æŒ‡å®šã—ã¦ãã ã•ã„"),o.createElement("div",{className:"article-header__properties"},o.createElement(ze,{title:"slug"},e.slug),o.createElement(ze,{title:"published"},"boolean"==typeof e.published?o.createElement(o.Fragment,null,e.published?"trueï¼ˆå…¬é–‹ï¼‰":"falseï¼ˆä¸‹æ›¸ãï¼‰"):"true ã‚‚ã—ãã¯ false ã‚’æŒ‡å®šã—ã¦ãã ã•ã„"),o.createElement(ze,{title:"type"},"tech"===e.type?"techï¼ˆæŠ€è¡“è¨˜äº‹ï¼‰":o.createElement(o.Fragment,null,"idea"===e.type?"ideaï¼ˆã‚¢ã‚¤ãƒ‡ã‚¢ï¼‰":"tech ã‚‚ã—ãã¯ idea ã‚’æŒ‡å®šã—ã¦ãã ã•ã„")),o.createElement(ze,{title:"topics"},Array.isArray(e.topics)&&e.topics.length?o.createElement($e,{topics:e.topics}):"æŒ‡å®šãŒå¿…è¦ã§ã™")),!!t.length&&o.createElement("div",{className:"article-header__validation-errors"},o.createElement(Me,{validationErrors:t}))))},Oe=n.header`
  padding: 3.3rem 0;
  background: var(--c-gray-bg);
  .article-header__emoji {
    margin-top: -0.5em;
    font-size: 56px;
    color: #000;
  }
  .article-header__title {
    font-size: 1.9rem;
  }
  .article-header__properties {
    margin-top: 1rem;
  }
  .article-header__validation-errors {
    margin-top: 1.4rem;
  }
`,Re=({slug:e})=>{const{data:t,error:r,isValidating:a,mutate:l}=z(`/api/articles/${e}`,{revalidateOnFocus:!1,revalidateOnReconnect:!1,errorRetryCount:3}),i=null==t?void 0:t.article;return H(`${(null==i?void 0:i.title)||e}ã®ãƒ—ãƒ¬ãƒ“ãƒ¥ãƒ¼`),A((()=>{l()})),i?o.createElement(o.Fragment,null,o.createElement(Se,{article:i}),o.createElement(U,null,o.createElement(Le,{className:"article-show"},o.createElement("div",{className:"article-show__content"},o.createElement(I,{rawHtml:i.bodyHtml||""}))))):a?o.createElement(Y,{margin:"5rem auto"}):o.createElement(w,{message:null==r?void 0:r.message})},Le=n.div`
  .article-show__content {
    padding: 3rem 0 18rem;
  }
`,Be=({book:e})=>{const t=m.exports.useMemo((()=>(e=>Ne(e,[te,re,ae,le,se,pe,de,ce,me,ue,ge,he,_e,Ee,be,fe,ye,xe]))(e)),[e]);return o.createElement(Ie,null,o.createElement(U,null,o.createElement("div",{className:"book-header__hero"},o.createElement("div",{className:"book-header__cover"},o.createElement("img",{src:e.coverDataUrl||"/static-images/book-cover.png",className:"book-header__cover-img",alt:"ã‚«ãƒãƒ¼ç”»åƒã®ãƒ—ãƒ¬ãƒ“ãƒ¥ãƒ¼",width:"160",height:"224"})),o.createElement("h1",{className:"book-header__title"},e.title||"titleã‚’æŒ‡å®šã—ã¦ãã ã•ã„")),o.createElement("div",{className:"book-header__properties"},o.createElement(ze,{title:"slug"},e.slug),o.createElement(ze,{title:"published"},"boolean"==typeof e.published?o.createElement(o.Fragment,null,e.published?"trueï¼ˆå…¬é–‹ï¼‰":"falseï¼ˆä¸‹æ›¸ãï¼‰"):"true ã‚‚ã—ãã¯ false ã‚’æŒ‡å®šã—ã¦ãã ã•ã„"),o.createElement(ze,{title:"price"},"number"==typeof e.price?e.price:"åŠè§’æ•°å­—ã§æŒ‡å®šã—ã¦ãã ã•ã„"),o.createElement(ze,{title:"topics"},Array.isArray(e.topics)&&e.topics.length?o.createElement($e,{topics:e.topics}):"æŒ‡å®šãŒå¿…è¦ã§ã™"),o.createElement(ze,{title:"summary"},o.createElement("div",{className:"book-header__summary"},e.summary||"æŒ‡å®šãŒå¿…è¦ã§ã™"))),!!t.length&&o.createElement("div",{className:"book-header__validation-errors"},o.createElement(Me,{validationErrors:t}))))},Ie=n.header`
  padding: 3.3rem 0;
  background: var(--c-gray-bg);
  .book-header__hero {
    display: flex;
    align-items: center;
  }
  .book-header__cover-img {
    background: #fff;
    object-fit: cover;
    box-shadow: 0 9px 20px -9px rgba(0, 27, 68, 0.52),
      0 0 3px rgba(0, 21, 60, 0.1);
    border-radius: 5px;
  }
  .book-header__title {
    flex: 1;
    margin-top: -1em;
    margin-left: 1.2em;
    font-size: 1.8rem;
  }
  .book-header__properties {
    margin-top: 1.5rem;
  }
  .book-header__summary {
    white-space: pre-line;
  }
  .book-header__validation-errors {
    margin-top: 1.4rem;
  }
  @media (max-width: 768px) {
    .book-header__hero {
      display: block;
    }
    .book-header__cover {
      text-align: center;
    }
    .book-header__title {
      margin: 1.7rem 0 0;
      font-size: 1.5rem;
    }
  }
`,Pe=({bookSlug:e,chapters:t,showListNumber:r})=>{const a=r?"ol":"ul";return o.createElement(Ue,{className:"book-chapter-list"},o.createElement(a,null,t.map((t=>o.createElement("li",{key:`book-chapter-${t.filename}`,className:"book-chapter-list__li"},o.createElement(at,{bookSlug:e,chapterFilename:t.filename,className:"book-chapter-list__link"},t.title||"ç„¡é¡Œ",o.createElement("span",{className:"book-chapter-list__slug"},"ï¼ˆ",t.slug||t.filename,"ï¼‰")))))))},Ue=n.div`
  color: var(--c-gray);
  font-weight: 700;
  ul,
  ol {
    margin-left: 1.4em;
    padding-left: 0.2em;
  }
  ul {
    list-style: disc;
  }
  ol {
    list-style: decimal;
  }
  .book-chapter-list__li {
    margin: 0.3rem 0;
  }
  .book-chapter-list__link {
    color: var(--c-primary-darker);
    font-weight: normal;
    &:hover {
      text-decoration: underline;
    }
  }
  .book-chapter-list__slug {
    font-size: 0.8em;
    opacity: 0.7;
    text-decoration: none;
  }
`,He=({slug:e})=>{const{data:t,error:r,isValidating:a,mutate:l}=z(`/api/books/${e}`,{revalidateOnFocus:!1,revalidateOnReconnect:!1,errorRetryCount:3}),i=null==t?void 0:t.book;H(`${(null==i?void 0:i.title)||e}ã®ãƒ—ãƒ¬ãƒ“ãƒ¥ãƒ¼`);const{data:n,error:s,isValidating:c,mutate:p}=z(`/api/books/${e}/chapters`,{revalidateOnFocus:!1,revalidateOnReconnect:!1,errorRetryCount:3});A((()=>{l(),p()}));const u=null==n?void 0:n.chapters,{withPositionChapters:g,withoutPositionChapters:h}=m.exports.useMemo((()=>u?u.reduce(((e,t)=>(e["number"==typeof t.position?"withPositionChapters":"withoutPositionChapters"].push(t),e)),{withPositionChapters:[],withoutPositionChapters:[]}):{withPositionChapters:[],withoutPositionChapters:[]}),[u]);return i?o.createElement(o.Fragment,null,o.createElement(Be,{book:i}),o.createElement(U,null,o.createElement(Te,{className:"book-show"},o.createElement("h2",{className:"book-show__chapters-title"},"Chapters"),c?o.createElement(Y,null):o.createElement(o.Fragment,null,s?o.createElement(w,{message:null==s?void 0:s.message}):o.createElement("div",null,!!(null==g?void 0:g.length)&&o.createElement("div",{className:"book-show__chapters"},o.createElement(Pe,{bookSlug:i.slug,chapters:g,showListNumber:!0})),!!(null==h?void 0:h.length)&&o.createElement("div",{className:"book-show__excluded-chapters"},o.createElement("p",{className:"book-show__chapters-note"},"ä»¥ä¸‹ã®ãƒãƒ£ãƒ—ã‚¿ãƒ¼ã¯zenn.devã¸ã®ãƒ‡ãƒ—ãƒ­ã‚¤å¯¾è±¡ã«å«ã¾ã‚Œã¾ã›ã‚“ã€‚æœ¬ã«å«ã‚ã‚‹ãƒãƒ£ãƒ—ã‚¿ãƒ¼ã®",o.createElement("a",{href:"https://zenn.dev/zenn/articles/what-is-slug",rel:"noopener noreferrer",target:"_blank"},"ã‚¹ãƒ©ãƒƒã‚°"),"ã‚’",o.createElement("code",null,"config.yaml"),"ã«æŒ‡å®šã™ã‚‹ã‹ã€å…¨ã¦ã®ãƒãƒ£ãƒ—ã‚¿ãƒ¼ã®ãƒ•ã‚¡ã‚¤ãƒ«åã‚’",o.createElement("code",null,"ãƒãƒ£ãƒ—ã‚¿ãƒ¼ç•ªå·.ã‚¹ãƒ©ãƒƒã‚°.md"),"ã®å½¢å¼ã«ã—ã¦ãã ã•ã„ã€‚",o.createElement("a",{href:"https://zenn.dev/zenn/articles/zenn-cli-guide#%F0%9F%93%84-%E5%90%84%E3%83%81%E3%83%A3%E3%83%97%E3%82%BF%E3%83%BC%E3%81%AE%E3%83%95%E3%82%A1%E3%82%A4%E3%83%AB%EF%BC%88%E2%97%AF%E2%97%AF.md%EF%BC%89",target:"_blank",rel:"noopener noreferrer"},"è©³ã—ã„èª¬æ˜Žã‚’é–‹ã",o.createElement(d,{className:"book-show__open-icon"}))),o.createElement("div",{className:"book-show__chapters"},o.createElement(Pe,{bookSlug:i.slug,chapters:h,showListNumber:!1})))))))):a?o.createElement(Y,{margin:"5rem auto"}):o.createElement(w,{message:null==r?void 0:r.message})},Te=n.div`
  padding: 3rem 0 6rem;

  .book-show__chapters-title {
    font-size: 2rem;
  }
  .book-show__chapters {
    margin-top: 1rem;
    line-height: 1.5;
  }
  .book-show__excluded-chapters {
    margin-top: 2.5rem;
    padding: 1.3rem;
    border: solid 1px var(--c-gray-border);
    border-radius: 10px;
    color: var(--c-gray);
    font-size: 0.95rem;
  }
  .book-show__chapters-note {
    line-height: 1.8;
    code {
      line-height: 1.2;
      display: inline-flex;
      margin: 0 0.2em;
      padding: 0.2em 0.5em;
      font-size: 0.88em;
      background: var(--c-gray-bg);
      border-radius: 5px;
    }
    a {
      display: inline-flex;
      align-items: center;
      text-decoration: underline;
      text-underline-offset: 4px;
      text-decoration-color: var(--c-gray-border);
      &:hover {
        color: var(--c-body);
      }
    }
  }
  .book-show__open-icon {
    width: 15px;
    height: 15px;
  }
`,Ke=({chapter:e,book:t})=>{const r=m.exports.useMemo((()=>(e=>Ne(e,[ke,re,ae,we]))(e)),[e]);return o.createElement(De,null,o.createElement("div",{className:"chapter-header__book"},o.createElement(rt,{slug:t.slug,className:"chapter-header__book-link"},o.createElement(E,{className:"book-header__book-back"}),o.createElement("span",{className:"chapter-header__book-title"},t.title))),o.createElement("div",{className:"chapter-header__main"},o.createElement(U,null,o.createElement("h1",{className:"chapter-header__title"},e.title||"titleã‚’æŒ‡å®šã—ã¦ãã ã•ã„"),o.createElement("div",{className:"chapter-header__properties"},o.createElement(ze,{title:"slug"},e.slug),o.createElement(ze,{title:"free"},0===t.price?"æœ¬ã®ä¾¡æ ¼ãŒÂ¥0ã§ã‚ã‚‹ãŸã‚ãƒãƒ£ãƒ—ã‚¿ãƒ¼ã¯ç„¡æ–™å…¬é–‹ã•ã‚Œã¾ã™":o.createElement(o.Fragment,null,e.free?"trueï¼ˆç„¡æ–™å…¬é–‹ï¼‰":"falseï¼ˆè³¼å…¥è€…ã®ã¿é–²è¦§å¯ï¼‰"))),!!r.length&&o.createElement("div",{className:"chapter-header__validation-errors"},o.createElement(Me,{validationErrors:r})))))},De=n.header`
  background: var(--c-gray-bg);
  .chapter-header__book {
    background: rgba(158, 186, 203, 0.3);
    padding: 0.6rem 1rem;
  }
  .chapter-header__book-link {
    display: flex;
    align-items: center;
    color: var(--c-gray);
    &:hover {
      color: var(--c-body);
    }
  }
  .book-header__book-back {
    width: 14px;
    height: 14px;
  }

  .chapter-header__book-title {
    margin-left: 5px;
    flex: 1;
    font-size: 13.5px;
    font-weight: 700;
  }
  .chapter-header__main {
    padding: 2.2rem 0 2.8rem;
  }
  .chapter-header__title {
    font-size: 1.8rem;
  }
  .chapter-header__properties {
    margin-top: 1rem;
  }
  .chapter-header__validation-errors {
    margin-top: 1.4rem;
  }
`,qe=({bookSlug:e,chapterFilename:t})=>{const{data:r,error:a,isValidating:l,mutate:i}=z(`/api/books/${e}`,{revalidateOnFocus:!1,revalidateOnReconnect:!1,errorRetryCount:3}),n=null==r?void 0:r.book,{data:s,error:c,isValidating:m,mutate:d}=z(`/api/books/${e}/chapters/${t}`,{revalidateOnFocus:!1,revalidateOnReconnect:!1,errorRetryCount:3}),p=null==s?void 0:s.chapter;return H(`${(null==p?void 0:p.title)||t}ã®ãƒ—ãƒ¬ãƒ“ãƒ¥ãƒ¼`),A((()=>{i(),d()})),n?p?o.createElement(o.Fragment,null,o.createElement(Ke,{book:n,chapter:p}),o.createElement(U,null,o.createElement(We,{className:"book-show"},o.createElement("div",{className:"chapter-show__content"},o.createElement(I,{rawHtml:p.bodyHtml||""}))))):m?o.createElement(Y,{margin:"5rem auto"}):o.createElement(w,{message:(null==c?void 0:c.message)||`ãƒãƒ£ãƒ—ã‚¿ãƒ¼ ${t} ã®ãƒ‡ãƒ¼ã‚¿ã‚’å–å¾—ã§ãã¾ã›ã‚“ã§ã—ãŸ`}):l?o.createElement(Y,{margin:"5rem auto"}):o.createElement(w,{message:(null==a?void 0:a.message)||`æœ¬ ${e} ã®ãƒ‡ãƒ¼ã‚¿ã‚’å–å¾—ã§ãã¾ã›ã‚“ã§ã—ãŸ`})},We=n.div`
  .chapter-show__content {
    padding: 3rem 0 10rem;
  }
`;const Ye=s.Path().any("slug",{action:({slug:e})=>o.createElement(Re,{slug:e})}),Ze=s.Path().any("slug",{action:({slug:e})=>o.createElement(He,{slug:e})}),Je=Ze.anyRoute.attach(s.Path()).any("chapter_filename",{action:({slug:e,chapter_filename:t})=>{return o.createElement(qe,{bookSlug:e,chapterFilename:(r=t,unescape(r).replace(/%2E/g,"."))});var r}}),Xe=s.SingleHash("hash",{optional:!0}).attach(s.Path()).any("slug",{action:({slug:e,hash:t})=>o.createElement(G,{slug:e,hash:t})}),Ge=s.Path().exact({action:()=>o.createElement(q,null)}).route("guide",(e=>e.attach(Xe))).route("articles",(e=>e.attach(Ye))).route("books",(e=>e.attach(Je))),Qe=()=>s.useRoutes(Ge),et=e=>{const t=s.useLocation(),r=m.exports.useMemo((()=>t.key===Ge.exactRoute.key),[t]);return o.createElement(s.Link,{route:Ge.exactRoute,className:r?`active ${e.className}`:e.className},e.children)},tt=e=>{const t=s.useLocation(),r=m.exports.useMemo((()=>t.pathname===`/articles/${e.slug}`),[t]);return o.createElement(s.Link,{route:Ye.anyRoute,match:{slug:e.slug},className:r?`active ${e.className}`:e.className},e.children)},rt=e=>{const t=s.useLocation(),r=m.exports.useMemo((()=>t.pathname===`/books/${e.slug}`),[t]);return o.createElement(s.Link,{route:Ze.anyRoute,match:{slug:e.slug},className:r?`active ${e.className}`:e.className},e.children)},at=e=>{const t=s.useLocation(),r=(a=e.chapterFilename,escape(a.replace(/\./g,"%2E")));var a;const l=m.exports.useMemo((()=>t.pathname===`/books/${e.bookSlug}/${r}`),[t]);return o.createElement(s.Link,{route:Je.anyRoute,match:{slug:e.bookSlug,chapter_filename:r},className:l?`active ${e.className}`:e.className},e.children)},lt=e=>{const t=s.useLocation(),r=m.exports.useMemo((()=>t.pathname===`/guide/${e.slug}`&&(!e.hash||t.hash===`#${e.hash}`)),[t]);return o.createElement(s.Link,{route:Xe.anyRoute,match:{slug:e.slug,hash:e.hash},className:r?`active ${e.className}`:e.className},e.children)},it=({article:e})=>o.createElement(tt,{slug:e.slug},o.createElement(V,{title:e.title||e.slug,label:e.published?void 0:"ä¸‹æ›¸ã",emoji:e.emoji||"ðŸ“„"})),nt=({bookSlug:e,chapter:t,chapterNumber:r})=>{const a="number"==typeof r;return o.createElement(at,{bookSlug:e,chapterFilename:t.filename},o.createElement(V,{label:a?void 0:"é™¤å¤–",title:`${a?`${r}. `:""}${t.title||t.slug}`}))},ot=({bookSlug:e})=>{const{data:t,mutate:r}=z(`/api/books/${e}/chapters`,{revalidateOnFocus:!1}),a=null==t?void 0:t.chapters;return A((()=>{r()})),o.createElement("ul",null,o.createElement("li",null,o.createElement(rt,{slug:e},o.createElement(V,{title:"è¨­å®š",emoji:"ðŸ“˜"}))),!!(null==a?void 0:a.length)&&o.createElement(o.Fragment,null,null==a?void 0:a.map(((t,r)=>o.createElement("li",{key:`dir-book-${e}-${t.slug}`},o.createElement(nt,{bookSlug:e,chapter:t,chapterNumber:"number"==typeof t.position?r+1:void 0}))))))},st=({book:e})=>o.createElement(O,{title:e.title||e.slug,uniqueKey:`dir-book-${e.slug}`,defaultOpen:!1,label:e.published?void 0:"ä¸‹æ›¸ã"},o.createElement(ot,{bookSlug:e.slug})),ct=({sort:e})=>{const{data:t,mutate:r}=z(`/api/articles?sort=${e}`,{revalidateOnFocus:!1,errorRetryCount:3}),a=null==t?void 0:t.articles;return A((()=>{r()})),o.createElement(O,{title:"articles",uniqueKey:"dir-articles",defaultOpen:!0},o.createElement("ul",null,!!(null==a?void 0:a.length)&&o.createElement("li",null,a.map((e=>o.createElement(it,{key:`item-article-${e.slug}`,article:e}))))))},mt=({sort:e})=>{const{data:t,mutate:r}=z(`/api/books?sort=${e}`,{revalidateOnFocus:!1}),a=null==t?void 0:t.books;return A((()=>{r()})),o.createElement(O,{title:"books",uniqueKey:"dir-books",defaultOpen:!0},!!(null==a?void 0:a.length)&&o.createElement("ul",null,null==a?void 0:a.map((e=>o.createElement(st,{key:`item-book-${e.slug}`,book:e})))))},dt=()=>{const[e,t]=M({cacheKey:"fold-sidebar",defaultValue:!1}),[r,a]=M({cacheKey:"item-sort-type",defaultValue:"modified"});return o.createElement(pt,{"aria-expanded":!e,className:"sidebar"},o.createElement("button",{className:"sidebar__btn-fold",onClick:()=>t(!e),"aria-label":e?"ãƒ¡ãƒ‹ãƒ¥ãƒ¼ã‚’é–‹ã":"æŠ˜ã‚ŠãŸãŸã‚€"},e?o.createElement(b,{className:"sidebar__fold-icon"}):o.createElement(v,{className:"sidebar__fold-icon"})),o.createElement("div",{className:"sidebar__inner","aria-hidden":e},o.createElement("header",{className:"sidebar__header"},o.createElement(et,null,o.createElement("img",{src:"/logo.svg",alt:"Zenn Editor",width:150,height:20,className:"sidebar__header-logo"})),o.createElement(L,{openButtonIcon:o.createElement(f,{className:"sidebar__sort-open"}),openButtonAriaLabel:"ã‚½ãƒ¼ãƒˆè¨­å®šã‚’é–‹ã",position:"right",options:[{value:"modified",label:"ãƒ•ã‚¡ã‚¤ãƒ«æ›´æ–°é †ã«ä¸¦ã¹ã‚‹"},{value:"system",label:"ã‚·ã‚¹ãƒ†ãƒ ã®è¡¨ç¤ºé †ã«å¾“ã†"}],value:r,setValue:e=>a(e),width:200})),o.createElement("div",{className:"sidebar__items"},o.createElement(ct,{sort:r}),o.createElement(mt,{sort:r}),o.createElement("ul",{className:"sidebar__static-links"},o.createElement("li",null,o.createElement(lt,{slug:"zenn-cli-guide",hash:"cli-%E3%81%A7%E8%A8%98%E4%BA%8B%EF%BC%88article%EF%BC%89%E3%82%92%E7%AE%A1%E7%90%86%E3%81%99%E3%82%8B"},o.createElement(V,{title:"è¨˜äº‹ã®ä½œæˆã‚¬ã‚¤ãƒ‰",emoji:"ðŸ“"}))),o.createElement("li",null,o.createElement(lt,{slug:"zenn-cli-guide",hash:"cli-%E3%81%A7%E6%9C%AC%EF%BC%88book%EF%BC%89%E3%82%92%E7%AE%A1%E7%90%86%E3%81%99%E3%82%8B"},o.createElement(V,{title:"æœ¬ã®ä½œæˆã‚¬ã‚¤ãƒ‰",emoji:"ðŸ“š"}))),o.createElement("li",null,o.createElement(lt,{slug:"deploy-github-images"},o.createElement(V,{title:"ç”»åƒç®¡ç†ã‚¬ã‚¤ãƒ‰",emoji:"ðŸž",label:"Beta"}))),o.createElement("li",null,o.createElement("a",{href:"https://zenn.dev/zenn/articles/markdown-guide",target:"_blank",rel:"noopener noreferrer"},o.createElement(V,{title:"ãƒžãƒ¼ã‚¯ãƒ€ã‚¦ãƒ³è¨˜æ³•",emoji:"ðŸ–‹ï¸ ",showNewTabIcon:!0}))),o.createElement("li",null,o.createElement("a",{href:"https://zenn.dev/dashboard/uploader",target:"_blank",rel:"noopener noreferrer"},o.createElement(V,{title:"ç”»åƒã®ã‚¢ãƒƒãƒ—ãƒ­ãƒ¼ãƒ‰",emoji:"ðŸ“·",showNewTabIcon:!0})))))))},pt=n.div`
  position: sticky;
  top: 0;
  height: 100vh;
  overflow-y: auto;
  border-right: solid 1px var(--c-gray-border);
  padding: 15px;
  width: 46px;
  &[aria-expanded='true'] {
    width: 350px;
  }
  .sidebar__btn-fold {
    position: absolute;
    top: 12px;
    right: 10px;
    display: inline-flex;
    align-items: center;
    justify-content: center;
    width: 26px;
    height: 26px;
    background: var(--c-gray-bg);
    border-radius: 5px;
    &:hover {
      .sidebar__fold-icon {
        color: var(--c-body);
      }
    }
  }
  .sidebar__fold-icon {
    width: 18px;
    width: 18px;
    color: var(--c-gray);
  }

  .sidebar__inner {
    &[aria-hidden='true'] {
      display: none;
    }
  }
  .sidebar__header {
    display: flex;
    align-items: center;
    justify-content: space-between;
    padding-right: 34px;
  }
  .sidebar__header-logo {
    flex-shrink: 0;
    display: block;
  }
  .sidebar__sort-open {
    width: 22px;
    height: 22px;
  }
  .sidebar__items {
    margin: 10px 0;
    display: block;
    font-size: 13.5px;
    a {
      margin: 2px 0;
      padding: 2px 0;
      display: block;
      color: var(--c-gray);
      &:hover {
        color: var(--c-body);
      }
      &.active {
        position: relative;
        color: var(--c-primary-darker);
        &:before {
          content: '';
          position: absolute;
          background: var(--c-primary-bg);
          top: 0;
          bottom: 0;
          left: -5px;
          right: -5px;
          z-index: -1;
          border-radius: 4px;
        }
      }
    }
  }
  .sidebar__static-links {
    margin-top: 8px;
  }
`,ut=({children:e})=>o.createElement(gt,{className:"layout"},o.createElement("aside",{className:"layout__sidebar"},o.createElement(dt,null)),o.createElement("main",{className:"layout__main"},e)),gt=n.div`
  display: flex;
  .layout__main {
    flex: 1;
  }
`;const ht=()=>{const e=s.useLocation();return m.exports.useEffect((()=>{e.hash||window.scrollTo(0,0)}),[e]),null},_t=()=>(m.exports.useEffect((()=>{!function(e,t){if(!t)return e();if(void 0===x){const e=document.createElement("link").relList;x=e&&e.supports&&e.supports("modulepreload")?"modulepreload":"preload"}Promise.all(t.map((e=>{if(e in k)return;k[e]=!0;const t=e.endsWith(".css"),r=t?'[rel="stylesheet"]':"";if(document.querySelector(`link[href="${e}"]${r}`))return;const a=document.createElement("link");return a.rel=t?"stylesheet":x,t||(a.as="script",a.crossOrigin=""),a.href=e,document.head.appendChild(a),t?new Promise(((e,t)=>{a.addEventListener("load",e),a.addEventListener("error",t)})):void 0}))).then((()=>e()))}((()=>import("./index.4ffda0fe.js")),void 0)}),[]),o.createElement($,null,o.createElement(j,null,o.createElement(s.RoconRoot,null,o.createElement(ht,null),o.createElement(ut,null,o.createElement(Qe,null))))));y.render(o.createElement(_t,null),document.getElementById("root"));
