// zenn-markdown-html/lib/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var markdown_it_1 = __importDefault(require("markdown-it"));
var crypto_1 = __importDefault(require("crypto"));
// plugis
var md_container_1 = require("./utils/md-container");
var md_renderer_fence_1 = require("./utils/md-renderer-fence");
var md_linkify_to_card_1 = require("./utils/md-linkify-to-card");
var md_katex_1 = require("./utils/md-katex");
var md_br_1 = require("./utils/md-br");
var md_custom_block_1 = require("./utils/md-custom-block");
var markdown_it_imsize_1 = __importDefault(require("@steelydylan/markdown-it-imsize"));
var markdown_it_anchor_1 = __importDefault(require("markdown-it-anchor"));
var mdContainer = require('markdown-it-container');
var mdFootnote = require('markdown-it-footnote');
var mdTaskLists = require('markdown-it-task-lists');
var mdInlineComments = require('markdown-it-inline-comments');
var mdLinkAttributes = require('markdown-it-link-attributes');
var md = (0, markdown_it_1.default)({
    breaks: true,
    linkify: true,
});
md.linkify.set({ fuzzyLink: false });
md.use(md_br_1.mdBr)
    .use(md_renderer_fence_1.mdRendererFence)
    .use(markdown_it_imsize_1.default)
    .use(md_custom_block_1.mdCustomBlock)
    .use(mdContainer, 'details', md_container_1.containerDetailsOptions)
    .use(mdContainer, 'message', md_container_1.containerMessageOptions)
    .use(mdFootnote)
    .use(mdTaskLists, { enabled: true })
    .use(mdInlineComments)
    .use(mdLinkAttributes, {
    pattern: /^(?!https:\/\/zenn\.dev\/)/,
    attrs: {
        rel: 'nofollow',
    },
})
    .use(markdown_it_anchor_1.default, {
    level: [1, 2, 3, 4],
    permalink: markdown_it_anchor_1.default.permalink.ariaHidden({
        placement: 'before',
        class: 'header-anchor-link',
        symbol: '',
    }),
    tabIndex: false,
})
    .use(md_katex_1.mdKatex)
    .use(md_linkify_to_card_1.mdLinkifyToCard);
// custom footnote
md.renderer.rules.footnote_block_open = function () {
    return '<section class="footnotes">\n' +
        '<div class="footnotes-title">脚注</div>\n' +
        '<ol class="footnotes-list">\n';
};
var markdownToHtml = function (text) {
    if (!(text && text.length))
        return '';
    // docIdは複数のコメントが1ページに指定されたときに脚注のリンク先が重複しないように指定する
    // 1ページの中で重複しなければ問題ないため、ごく短いランダムな文字列とする
    // - https://github.com/zenn-dev/zenn-community/issues/356
    // - https://github.com/markdown-it/markdown-it-footnote/pull/8
    var docId = crypto_1.default.randomBytes(2).toString('hex');
    return md.render(text, { docId: docId });
};
exports.default = markdownToHtml;
